### PR TITLE
xds: moved shared matchers to internal/xds

### PR DESCRIPTION
**Pull request by zasweq**
This PR moves the header matcher previously present in xds_resolver to internal/xds/matchers. It also moves the string matcher previously present as a file in internal/xds to this new package internal/xds/matchers.

### CODE REVIEW

File: internal/xds/matcher/matcher_header.go [hunk 1]
```diff
@@ -0,0 +1,224 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
```
Comment (menghanl) [inline internal/xds/matcher/matcher_header.go:3]: 2021

File: internal/xds/matcher/string_matcher.go [hunk 2]
```diff
@@ -16,9 +16,9 @@
  *
  */
 
-// Package xds contains types that need to be shared between code under
+// Package matcher contains types that need to be shared between code under
```
Comment (menghanl) [inline internal/xds/matcher/string_matcher.go:19]: contains _matchers_?

**Review by menghanl — APPROVED** [review]

File: internal/xds/matcher/matcher_header.go (hunk 1 above)
Comment (zasweq) [reply internal/xds/matcher/matcher_header.go:3]: Changed. (Will come in RBAC PR)

File: internal/xds/matcher/string_matcher.go (hunk 2 above)
Comment (zasweq) [reply internal/xds/matcher/string_matcher.go:19]: Changed. (Will come in RBAC PR)

**Pull request merged**